src/shogun/machine/TypedMachine.h

```diff
@@ -0,0 +1,70 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Shubham Shukla
+ */
+
+#ifndef _TYPEDMACHINE_H__
+#define _TYPEDMACHINE_H__
+
+#include <shogun/lib/config.h>
+
+#include <shogun/base/progress.h>
+#include <shogun/lib/SGVector.h>
+#include <shogun/lib/common.h>
+#include <shogun/machine/LinearMachine.h>
+
+namespace shogun
+{
+	class CFeatures;
+	class CLabels;
+	
+	template <class P, class T>
+	class CTypedMachine : public T
+	{
+	public:
+		/** Default constructor */
+		CTypedMachine() : T()
+		{
+		}
+
+		virtual ~CTypedMachine()
+		{
+		}
+
+	protected:
+		virtual bool train_machine(CFeatures* data = NULL)
+		{
+				if (!data)
+				{	
+					REQUIRE(this->features, "No features provided.\n")	
```

Comment:
what is `this` here? This is inherited from the base class I guess?

Reply:
this would be CTypedMachine<CLARS, CLinearMachine>

Reply:
but whose features are this?
CLARS right? So this means that your implementation here depends on the fact that the first template parameter has a member called features. We dont want to enforce that actually.

Reply:
features is CLinearMachine member CLARS do not store them locally... so i think all will have that member like CLDA because it is in machine

Reply:
my point is that this implementation assumes that T or P has a member called features, i.e. it wont compile otherwise.
I would just enforce features to be passed here tbh.
We wanna to this anyways soon, so all the mixins can already require features to be passed, so they dont need to be members

Reply:
We can use something like static_assert for meaningful messages